esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -1241,194 +1241,184 @@ void setDeleteFileResults(const char* fileName, const char* nodeGroup, bool fail
-}
+                const char *lfn = dfsLFN.get();
+                Owned<IDistributedFile> df = fdir.lookup(lfn, userdesc, true, false, false, nullptr, defaultPrivilegedUser, 30000); // 30 sec timeout
+                if (!df)
+                {
+                    VStringBuffer message("Could not delete file %s: file not found.", lfn);
+                    addResult(lfn, group, true, message);
+                    continue;
+                }
 
-bool CWsDfuEx::DFUDeleteFiles(IEspContext &context, IEspDFUArrayActionRequest &req, IEspDFUArrayActionResponse &resp)
-{
-    if (isDetachedFromDali())
-        throw MakeStringException(ECLWATCH_INVALID_INPUT, "ESP server is detached from Dali. Please try later.");
+                Owned<IDistributedSuperFileIterator> superOwners = df->getOwningSuperFiles();
+                if (!removeFromSuperFiles && superOwners->first())
+                {
+                    VStringBuffer message("The file %s can not be deleted because it is owned by one or more superfiles.", lfn);
+                    addResult(lfn, group, true, message);
+                    continue;
+                }
 
-    double version = context.getClientVersion();
-    Owned<IUserDescriptor> userdesc;
-    const char *username = context.queryUserId();
-    if(username && *username)
-    {
-        userdesc.setown(createUserDescriptor());
-        userdesc->set(username, context.queryPassword(), context.querySignature());
-    }
+                ForEach(*superOwners)
+                {
+                    IDistributedSuperFile &superOwner = superOwners->query();
+                    try
+                    {
+                        superOwner.removeSubFile(lfn, false, false, nullptr);
+                        PROGLOG("File %s is removed from superfile %s", lfn, superOwner.queryLogicalName());
+                    }
+                    catch(IException *e)
+                    {
+                        VStringBuffer message("Could not delete file %s from superfile %s: ", lfn, superOwner.queryLogicalName());
+                        e->errorMessage(message);
+                        addResult(lfn, group, true, message);
+                        e->Release();
+                        break;
+                    }
+                    catch(...)
+                    {
+                        VStringBuffer message("Could not delete file %s from superfile %s: unknown exception", lfn, superOwner.queryLogicalName());
+                        addResult(lfn, group, true, message);
+                        break;
+                    }
 
-    StringBuffer returnStr;
-    StringBuffer auditStr(",FileAccess,WsDfu,DELETED,");
-    IArrayOf<IEspDFUActionInfo> actionResults;
+                    if (removeEmptyOwningSuperFiles && (superOwner.numSubFiles(false)==0) && !logicalFileNames.contains(superOwner.queryLogicalName()))
```

Comment:
In fact, the whole logic here may not correct. For example, a super file super_a contains only one subfile a and the super_a has a superOwner super_s_a. After we remove the subfile a from the super_a, the super_a has 0 subfile. If the super_a is added to the emptyOwningSuperFiles, it will cause an error when the files in the emptyOwningSuperFiles are deleted.

Reply:
Should I add a recursive check/remove/add function to work from a superOwner and another superOwner? Will it cause some locks by the existing IDistributedFile/IDistributedSuperFile ?

Reply:
@wangkx - you're right, that's an existing bug though, and this refactoring to make this service more efficient/avoid spurious logging etc., doesn't change the fact it didn't handle deletions of empty nested supers as far as I can tell?

If so, I'd think about fixing it, but not in this PR.

Reply:
@jakesmith I just did the fix. Please take a look. If you want, I can roll back.